File: app/src/main/java/de/xikolo/utils/DownloadUtil.java
package de.xikolo.utils;

import java.io.File;

import de.xikolo.models.Course;
import de.xikolo.models.Item;
import de.xikolo.models.Section;
import de.xikolo.models.Video;

public class DownloadUtil {

    public static final String TAG = DownloadUtil.class.getSimpleName();

    public enum VideoAssetType {
        SLIDES, TRANSCRIPT, VIDEO_SD, VIDEO_HD, AUDIO;

        public String getFileSuffix() {
            switch (this) {
                case SLIDES:
                    return "_slides.pdf";
                case TRANSCRIPT:
                    return "_transcript.pdf";
                case VIDEO_SD:
                    return "_video_sd.mp4";
                case VIDEO_HD:
                    return "_video_hd.mp4";
                case AUDIO:
                    return "_audio.mp3";
                default:
                    return "";
            }
        }

        @Override
        public String toString() {
            switch (this) {
                case SLIDES:
                    return "Slides";
                case TRANSCRIPT:
                    return "Transcript";
                case VIDEO_SD:
                    return "SD Video";
                case VIDEO_HD:
                    return "HD Video";
                case AUDIO:
                    return "Audio";
                default:
                    return "";
            }
        }
    }

    public static String getVideoAssetFilePath(VideoAssetType type, Course course, Section section, Item item) {
        return FileUtil.createPublicAppFolderPath() + File.separator
                + FileUtil.escapeFilename(course.title) + "_" + course.id + File.separator
                + FileUtil.escapeFilename(section.title) + "_" + section.id + File.separator
                + FileUtil.escapeFilename(item.title) + type.getFileSuffix();
    }

    public static String getVideoAssetUrl(VideoAssetType type, Video video) {
        switch (type) {
            case VIDEO_HD:
                return video.singleStream.hdUrl;
            case VIDEO_SD:
                return video.singleStream.sdUrl;
            case SLIDES:
                return video.slidesUrl;
            case TRANSCRIPT:
                return video.transcriptUrl;
            case AUDIO:
                return video.audioUrl;
            default:
                return null;
        }
    }

}

File: app/src/main/java/de/xikolo/utils/DownloadUtil.kt
package de.xikolo.utils

import java.io.File

object DownloadUtil {

    val TAG: String = DownloadUtil::class.java.simpleName

    // the file path should identify a download uniquely, thus this AssetDownload object can be used as an identifier for downloads
    sealed class AssetDownload(val url: String?, open val fileName: String) {
        protected open val fileFolder: String = FileUtil.createPublicAppFolderPath() // must not end with separator

        open val title: String
            get() = fileName

        val filePath: String
            get() = fileFolder + File.separator + fileName


        sealed class Course(url: String?, override val fileName: String, val course: de.xikolo.models.Course) : AssetDownload(url, fileName) {
            override val fileFolder = super.fileFolder + File.separator + FileUtil.escapeFilename(course.title) + "_" + FileUtil.escapeFilename(course.id)

            sealed class Item constructor(url: String?, fileName: String, val item: de.xikolo.models.Item, val video: de.xikolo.models.Video) : Course(url, fileName, item.section.course) {
                override val fileFolder = super.fileFolder + File.separator + FileUtil.escapeFilename(item.section.title) + "_" + item.section.id
                override val fileName = FileUtil.escapeFilename(item.title) + "_" + fileName
                abstract val size: Int

                class Slides(item: de.xikolo.models.Item, video: de.xikolo.models.Video) : Item(video.slidesUrl, "slides.pdf", item, video) {
                    override val title = "Slides \"" + item.title + "\""
                    override val size = video.slidesSize
                }

                class Transcript(item: de.xikolo.models.Item, video: de.xikolo.models.Video) : Item(video.transcriptUrl, "transcript.pdf", item, video) {
                    override val title = "Transcript \"" + item.title + "\""
                    override val size = video.transcriptSize
                }

                class VideoSD(item: de.xikolo.models.Item, video: de.xikolo.models.Video) : Item(video.singleStream.sdUrl, "video_sd.mp4", item, video) {
                    override val title = "SD Video \"" + item.title + "\""
                    override val size = video.singleStream.sdSize
                }

                class VideoHD(item: de.xikolo.models.Item, video: de.xikolo.models.Video) : Item(video.singleStream.hdUrl, "video_hd.mp4", item, video) {
                    override val title = "HD Video \"" + item.title + "\""
                    override val size = video.singleStream.hdSize
                }

                class Audio(item: de.xikolo.models.Item, video: de.xikolo.models.Video) : Item(video.audioUrl, "audio.mp3", item, video) {
                    override val title = "Audio \"" + item.title + "\""
                    override val size = video.transcriptSize
                }
            }
        }
    }
}
